File: captainhcg/go/101-symmetric-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil{
        return true
    }
    return symHelper(root.Left, root.Right)
}

func symHelper(left *TreeNode, right *TreeNode) bool {
    if left == nil && right == nil {
        return true
    }
    if left == nil || right == nil {
        return false
    }
    if left.Val != right.Val {
        return false
    }
    return symHelper(left.Left, right.Right) && symHelper(left.Right, right.Left)
}

File: bwu/binary_tree/101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
            
        return self.isMirror(root.left, root.right)
        
    def isMirror(self, root1, root2):
        if root1 is None and root2 is None:
            return True
        elif root1 is None or root2 is None:
            return False
            
        return root1.val == root2.val \
            and self.isMirror(root1.left, root2.right) \
            and self.isMirror(root1.right, root2.left)
